diff --git a/Makefile.release b/Makefile.release
index 28570aa2..fc68c7ec 100644
--- a/Makefile.release
+++ b/Makefile.release
@@ -14,7 +14,7 @@ define MEDPRSTMT
 from pkg_resources import resource_filename as rfn;
 print(rfn('metapensiero.extjs.desktop',''))
 endef
-MEDSRC := $(shell $(PYTHON) -c "$(MEDPRSTMT)")
+MEDSRC := @medsrc@
 PACKED_CSS := $(SOLSRC)/static/all-styles.css
 PACKED_JS := $(SOLSRC)/static/all-classes.js
 MINIFIER := $(BINDIR)minify_js_scripts \
